File: packages/dartcv/lib/src/contrib/aruco.dart
// Copyright (c) 2024, rainyl and all contributors. All rights reserved.
// Use of this source code is governed by a Apache-2.0 license
// that can be found in the LICENSE file.

library cv.contrib;

import 'dart:ffi' as ffi;

import 'package:ffi/ffi.dart';

import '../core/base.dart';
import '../core/mat.dart';
import '../core/point.dart';
import '../core/scalar.dart';
import '../core/vec.dart';
import '../g/contrib.g.dart' as cvg;
import '../native_lib.dart' show ccontrib;
import 'aruco_dict.dart';

class ArucoDetector extends CvStruct<cvg.ArucoDetector> {
  ArucoDetector._(cvg.ArucoDetectorPtr ptr, [bool attach = true]) : super.fromPointer(ptr) {
    if (attach) {
      finalizer.attach(this, ptr.cast(), detach: this);
    }
  }
  factory ArucoDetector.fromPointer(
    cvg.ArucoDetectorPtr ptr, [
    bool attach = true,
  ]) =>
      ArucoDetector._(ptr, attach);

  factory ArucoDetector.empty() {
    final p = calloc<cvg.ArucoDetector>();
    cvRun(() => ccontrib.cv_aruco_arucoDetector_create(p));
    return ArucoDetector._(p);
  }

  factory ArucoDetector.create(ArucoDictionary dictionary, ArucoDetectorParameters parameters) {
    final p = calloc<cvg.ArucoDetector>();
    cvRun(
      () => ccontrib.cv_aruco_arucoDetector_create_1(
        dictionary.ref,
        parameters.ref,
        p,
      ),
    );
    return ArucoDetector.fromPointer(p);
  }

  @override
  cvg.ArucoDetector get ref => ptr.ref;
  static final finalizer =
      OcvFinalizer<cvg.ArucoDetectorPtr>(ccontrib.addresses.cv_aruco_arucoDetector_close);

  void dispose() {
    finalizer.detach(this);
    ccontrib.cv_aruco_arucoDetector_close(ptr);
  }

  /// DetectMarkers does basic marker detection.
  ///
  /// For further details, please see:
  /// https://docs.opencv.org/master/d9/d6a/group__aruco.html#ga3bc50d61fe4db7bce8d26d56b5a6428a
  (VecVecPoint2f corners, VecI32 ids, VecVecPoint2f rejectedImgPoints) detectMarkers(InputArray image) {
    final corners = VecVecPoint2f();
    final rejected = VecVecPoint2f();
    final ids = VecI32();
    cvRun(
      () => ccontrib.cv_aruco_arucoDetector_detectMarkers(
        ref,
        image.ref,
        corners.ptr,
        ids.ptr,
        rejected.ptr,
        ffi.nullptr,
      ),
    );
    return (corners, ids, rejected);
  }
}

void arucoDrawDetectedMarkers(
  Mat img,
  VecVecPoint2f markerCorners,
  VecI32 markerIds,
  Scalar borderColor,
) {
  cvRun(
    () => ccontrib.cv_aruco_drawDetectedMarkers(
      img.ref,
      markerCorners.ref,
      markerIds.ref,
      borderColor.ref,
      ffi.nullptr,
    ),
  );
}

Mat arucoGenerateImageMarker(
  PredefinedDictionaryType dictionaryId,
  int id,
  int sidePixels,
  int borderBits, [
  Mat? outImg,
]) {
  outImg ??= Mat.empty();
  cvRun(
    () => ccontrib.cv_aruco_generateImageMarker(
      dictionaryId.value,
      id,
      sidePixels,
      borderBits,
      outImg!.ref,
      ffi.nullptr,
    ),
  );
  return outImg;
}

class ArucoDetectorParameters extends CvStruct<cvg.ArucoDetectorParams> {
  ArucoDetectorParameters._(
    cvg.ArucoDetectorParamsPtr ptr, [
    bool attach = true,
  ]) : super.fromPointer(ptr) {
    if (attach) {
      finalizer.attach(this, ptr.cast(), detach: this);
    }
  }
  factory ArucoDetectorParameters.fromPointer(
    cvg.ArucoDetectorParamsPtr ptr, [
    bool attach = true,
  ]) =>
      ArucoDetectorParameters._(ptr, attach);

  factory ArucoDetectorParameters.empty() {
    final p = calloc<cvg.ArucoDetectorParams>();
    cvRun(() => ccontrib.cv_aruco_detectorParameters_create(p));
    return ArucoDetectorParameters._(p);
  }

  @override
  cvg.ArucoDetectorParams get ref => ptr.ref;
  static final finalizer = OcvFinalizer<cvg.ArucoDetectorParamsPtr>(
    ccontrib.addresses.cv_aruco_detectorParameters_close,
  );

  void dispose() {
    finalizer.detach(this);
    ccontrib.cv_aruco_detectorParameters_close(ptr);
  }

  int get adaptiveThreshWinSizeMin => ccontrib.cv_aruco_detectorParameters_get_adaptiveThreshWinSizeMin(ref);

  set adaptiveThreshWinSizeMin(int value) =>
      ccontrib.cv_aruco_detectorParameters_set_adaptiveThreshWinSizeMin(ref, value);

  int get adaptiveThreshWinSizeMax => ccontrib.cv_aruco_detectorParameters_get_adaptiveThreshWinSizeMax(ref);

  set adaptiveThreshWinSizeMax(int value) =>
      ccontrib.cv_aruco_detectorParameters_set_adaptiveThreshWinSizeMax(ref, value);

  int get adaptiveThreshWinSizeStep =>
      ccontrib.cv_aruco_detectorParameters_get_adaptiveThreshWinSizeStep(ref);

  set adaptiveThreshWinSizeStep(int value) =>
      ccontrib.cv_aruco_detectorParameters_set_adaptiveThreshWinSizeStep(ref, value);

  double get adaptiveThreshConstant => ccontrib.cv_aruco_detectorParameters_get_adaptiveThreshConstant(ref);

  set adaptiveThreshConstant(double value) =>
      ccontrib.cv_aruco_detectorParameters_set_adaptiveThreshConstant(ref, value);

  double get minMarkerPerimeterRate => ccontrib.cv_aruco_detectorParameters_get_minMarkerPerimeterRate(ref);

  set minMarkerPerimeterRate(double value) =>
      ccontrib.cv_aruco_detectorParameters_set_minMarkerPerimeterRate(ref, value);

  double get maxMarkerPerimeterRate => ccontrib.cv_aruco_detectorParameters_get_maxMarkerPerimeterRate(ref);

  set maxMarkerPerimeterRate(double value) =>
      ccontrib.cv_aruco_detectorParameters_set_maxMarkerPerimeterRate(ref, value);

  double get polygonalApproxAccuracyRate =>
      ccontrib.cv_aruco_detectorParameters_get_polygonalApproxAccuracyRate(ref);

  set polygonalApproxAccuracyRate(double value) =>
      ccontrib.cv_aruco_detectorParameters_set_polygonalApproxAccuracyRate(ref, value);

  double get minCornerDistanceRate => ccontrib.cv_aruco_detectorParameters_get_minCornerDistanceRate(ref);

  set minCornerDistanceRate(double value) =>
      ccontrib.cv_aruco_detectorParameters_set_minCornerDistanceRate(ref, value);

  int get minDistanceToBorder => ccontrib.cv_aruco_detectorParameters_get_minDistanceToBorder(ref);

  set minDistanceToBorder(int value) =>
      ccontrib.cv_aruco_detectorParameters_set_minDistanceToBorder(ref, value);

  double get minMarkerDistanceRate => ccontrib.cv_aruco_detectorParameters_get_minMarkerDistanceRate(ref);

  set minMarkerDistanceRate(double value) =>
      ccontrib.cv_aruco_detectorParameters_set_minMarkerDistanceRate(ref, value);

  int get cornerRefinementMethod => ccontrib.cv_aruco_detectorParameters_get_cornerRefinementMethod(ref);

  set cornerRefinementMethod(int value) =>
      ccontrib.cv_aruco_detectorParameters_set_cornerRefinementMethod(ref, value);

  int get cornerRefinementWinSize => ccontrib.cv_aruco_detectorParameters_get_cornerRefinementWinSize(ref);

  set cornerRefinementWinSize(int value) =>
      ccontrib.cv_aruco_detectorParameters_set_cornerRefinementWinSize(ref, value);

  int get cornerRefinementMaxIterations =>
      ccontrib.cv_aruco_detectorParameters_get_cornerRefinementMaxIterations(ref);

  set cornerRefinementMaxIterations(int value) =>
      ccontrib.cv_aruco_detectorParameters_set_cornerRefinementMaxIterations(ref, value);

  double get cornerRefinementMinAccuracy =>
      ccontrib.cv_aruco_detectorParameters_get_cornerRefinementMinAccuracy(ref);

  set cornerRefinementMinAccuracy(double value) =>
      ccontrib.cv_aruco_detectorParameters_set_cornerRefinementMinAccuracy(ref, value);

  int get markerBorderBits => ccontrib.cv_aruco_detectorParameters_get_markerBorderBits(ref);

  set markerBorderBits(int value) => ccontrib.cv_aruco_detectorParameters_set_markerBorderBits(ref, value);

  int get perspectiveRemovePixelPerCell =>
      ccontrib.cv_aruco_detectorParameters_get_perspectiveRemovePixelPerCell(ref);

  set perspectiveRemovePixelPerCell(int value) =>
      ccontrib.cv_aruco_detectorParameters_set_perspectiveRemovePixelPerCell(ref, value);

  double get perspectiveRemoveIgnoredMarginPerCell =>
      ccontrib.cv_aruco_detectorParameters_get_perspectiveRemoveIgnoredMarginPerCell(ref);

  set perspectiveRemoveIgnoredMarginPerCell(double value) =>
      ccontrib.cv_aruco_detectorParameters_set_perspectiveRemoveIgnoredMarginPerCell(ref, value);

  double get maxErroneousBitsInBorderRate =>
      ccontrib.cv_aruco_detectorParameters_get_maxErroneousBitsInBorderRate(ref);

  set maxErroneousBitsInBorderRate(double value) =>
      ccontrib.cv_aruco_detectorParameters_set_maxErroneousBitsInBorderRate(ref, value);

  double get minOtsuStdDev => ccontrib.cv_aruco_detectorParameters_get_minOtsuStdDev(ref);

  set minOtsuStdDev(double value) => ccontrib.cv_aruco_detectorParameters_set_minOtsuStdDev(ref, value);

  double get errorCorrectionRate => ccontrib.cv_aruco_detectorParameters_get_errorCorrectionRate(ref);

  set errorCorrectionRate(double value) =>
      ccontrib.cv_aruco_detectorParameters_set_errorCorrectionRate(ref, value);

  double get aprilTagQuadDecimate => ccontrib.cv_aruco_detectorParameters_get_aprilTagQuadDecimate(ref);

  set aprilTagQuadDecimate(double value) =>
      ccontrib.cv_aruco_detectorParameters_set_aprilTagQuadDecimate(ref, value);

  double get aprilTagQuadSigma => ccontrib.cv_aruco_detectorParameters_get_aprilTagQuadSigma(ref);

  set aprilTagQuadSigma(double value) =>
      ccontrib.cv_aruco_detectorParameters_set_aprilTagQuadSigma(ref, value);

  int get aprilTagMinClusterPixels => ccontrib.cv_aruco_detectorParameters_get_aprilTagMinClusterPixels(ref);

  set aprilTagMinClusterPixels(int value) =>
      ccontrib.cv_aruco_detectorParameters_set_aprilTagMinClusterPixels(ref, value);

  int get aprilTagMaxNmaxima => ccontrib.cv_aruco_detectorParameters_get_aprilTagMaxNmaxima(ref);

  set aprilTagMaxNmaxima(int value) =>
      ccontrib.cv_aruco_detectorParameters_set_aprilTagMaxNmaxima(ref, value);

  double get aprilTagCriticalRad => ccontrib.cv_aruco_detectorParameters_get_aprilTagCriticalRad(ref);

  set aprilTagCriticalRad(double value) =>
      ccontrib.cv_aruco_detectorParameters_set_aprilTagCriticalRad(ref, value);

  double get aprilTagMaxLineFitMse => ccontrib.cv_aruco_detectorParameters_get_aprilTagMaxLineFitMse(ref);

  set aprilTagMaxLineFitMse(double value) =>
      ccontrib.cv_aruco_detectorParameters_set_aprilTagMaxLineFitMse(ref, value);

  int get aprilTagMinWhiteBlackDiff =>
      ccontrib.cv_aruco_detectorParameters_get_aprilTagMinWhiteBlackDiff(ref);

  set aprilTagMinWhiteBlackDiff(int value) =>
      ccontrib.cv_aruco_detectorParameters_set_aprilTagMinWhiteBlackDiff(ref, value);

  int get aprilTagDeglitch => ccontrib.cv_aruco_detectorParameters_get_aprilTagDeglitch(ref);

  set aprilTagDeglitch(int value) => ccontrib.cv_aruco_detectorParameters_set_aprilTagDeglitch(ref, value);

  bool get detectInvertedMarker => ccontrib.cv_aruco_detectorParameters_get_detectInvertedMarker(ref);

  set detectInvertedMarker(bool value) =>
      ccontrib.cv_aruco_detectorParameters_set_detectInvertedMarker(ref, value);
}

File: packages/dartcv/src/dartcv/contrib/aruco.cpp
/*
    This file is originally from gocv project
    Licensed: Apache 2.0 license. Copyright (c) 2017-2021 The Hybrid Group.

    Modified by Rainyl.
    Licensed: Apache 2.0 license. Copyright (c) 2024 Rainyl.
*/
#include "dartcv/contrib/aruco.h"
#include "dartcv/core/vec.hpp"

CvStatus* cv_aruco_detectorParameters_create(ArucoDetectorParams* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::aruco::DetectorParameters();
    END_WRAP
}

void cv_aruco_detectorParameters_close(ArucoDetectorParamsPtr ap) {
    CVD_FREE(ap);
}

void cv_aruco_detectorParameters_set_adaptiveThreshWinSizeMin(ArucoDetectorParams self, int value) {
    self.ptr->adaptiveThreshWinSizeMin = value;
}

int cv_aruco_detectorParameters_get_adaptiveThreshWinSizeMin(ArucoDetectorParams self) {
    return self.ptr->adaptiveThreshWinSizeMin;
}

void cv_aruco_detectorParameters_set_adaptiveThreshWinSizeMax(ArucoDetectorParams self, int value) {
    self.ptr->adaptiveThreshWinSizeMax = value;
}

int cv_aruco_detectorParameters_get_adaptiveThreshWinSizeMax(ArucoDetectorParams self) {
    return self.ptr->adaptiveThreshWinSizeMax;
}

void cv_aruco_detectorParameters_set_adaptiveThreshWinSizeStep(
    ArucoDetectorParams self, int value
) {
    self.ptr->adaptiveThreshWinSizeStep = value;
}

int cv_aruco_detectorParameters_get_adaptiveThreshWinSizeStep(ArucoDetectorParams self) {
    return self.ptr->adaptiveThreshWinSizeStep;
}

void cv_aruco_detectorParameters_set_adaptiveThreshConstant(
    ArucoDetectorParams self, double value
) {
    self.ptr->adaptiveThreshConstant = value;
}

double cv_aruco_detectorParameters_get_adaptiveThreshConstant(ArucoDetectorParams self) {
    return self.ptr->adaptiveThreshConstant;
}

void cv_aruco_detectorParameters_set_minMarkerPerimeterRate(
    ArucoDetectorParams self, double value
) {
    self.ptr->minMarkerPerimeterRate = value;
}

double cv_aruco_detectorParameters_get_minMarkerPerimeterRate(ArucoDetectorParams self) {
    return self.ptr->minMarkerPerimeterRate;
}

void cv_aruco_detectorParameters_set_maxMarkerPerimeterRate(
    ArucoDetectorParams self, double value
) {
    self.ptr->maxMarkerPerimeterRate = value;
}

double cv_aruco_detectorParameters_get_maxMarkerPerimeterRate(ArucoDetectorParams self) {
    return self.ptr->maxMarkerPerimeterRate;
}

void cv_aruco_detectorParameters_set_polygonalApproxAccuracyRate(
    ArucoDetectorParams self, double value
) {
    self.ptr->polygonalApproxAccuracyRate = value;
}

double cv_aruco_detectorParameters_get_polygonalApproxAccuracyRate(ArucoDetectorParams self) {
    return self.ptr->polygonalApproxAccuracyRate;
}

void cv_aruco_detectorParameters_set_minCornerDistanceRate(ArucoDetectorParams self, double value) {
    self.ptr->minCornerDistanceRate = value;
}

double cv_aruco_detectorParameters_get_minCornerDistanceRate(ArucoDetectorParams self) {
    return self.ptr->minCornerDistanceRate;
}

void cv_aruco_detectorParameters_set_minDistanceToBorder(ArucoDetectorParams self, int value) {
    self.ptr->minDistanceToBorder = value;
}

int cv_aruco_detectorParameters_get_minDistanceToBorder(ArucoDetectorParams self) {
    return self.ptr->minDistanceToBorder;
}

void cv_aruco_detectorParameters_set_minMarkerDistanceRate(ArucoDetectorParams self, double value) {
    self.ptr->minMarkerDistanceRate = value;
}

double cv_aruco_detectorParameters_get_minMarkerDistanceRate(ArucoDetectorParams self) {
    return self.ptr->minMarkerDistanceRate;
}

void cv_aruco_detectorParameters_set_cornerRefinementMethod(ArucoDetectorParams self, int value) {
    self.ptr->cornerRefinementMethod = value;
}

int cv_aruco_detectorParameters_get_cornerRefinementMethod(ArucoDetectorParams self) {
    return self.ptr->cornerRefinementMethod;
}

void cv_aruco_detectorParameters_set_cornerRefinementWinSize(ArucoDetectorParams self, int value) {
    self.ptr->cornerRefinementWinSize = value;
}

int cv_aruco_detectorParameters_get_cornerRefinementWinSize(ArucoDetectorParams self) {
    return self.ptr->cornerRefinementWinSize;
}

void cv_aruco_detectorParameters_set_cornerRefinementMaxIterations(
    ArucoDetectorParams self, int value
) {
    self.ptr->cornerRefinementMaxIterations = value;
}

int cv_aruco_detectorParameters_get_cornerRefinementMaxIterations(ArucoDetectorParams self) {
    return self.ptr->cornerRefinementMaxIterations;
}

void cv_aruco_detectorParameters_set_cornerRefinementMinAccuracy(
    ArucoDetectorParams self, double value
) {
    self.ptr->cornerRefinementMinAccuracy = value;
}

double cv_aruco_detectorParameters_get_cornerRefinementMinAccuracy(ArucoDetectorParams self) {
    return self.ptr->cornerRefinementMinAccuracy;
}

void cv_aruco_detectorParameters_set_markerBorderBits(ArucoDetectorParams self, int value) {
    self.ptr->markerBorderBits = value;
}

int cv_aruco_detectorParameters_get_markerBorderBits(ArucoDetectorParams self) {
    return self.ptr->markerBorderBits;
}

void cv_aruco_detectorParameters_set_perspectiveRemovePixelPerCell(
    ArucoDetectorParams self, int value
) {
    self.ptr->perspectiveRemovePixelPerCell = value;
}

int cv_aruco_detectorParameters_get_perspectiveRemovePixelPerCell(ArucoDetectorParams self) {
    return self.ptr->perspectiveRemovePixelPerCell;
}

void cv_aruco_detectorParameters_set_perspectiveRemoveIgnoredMarginPerCell(
    ArucoDetectorParams self, double value
) {
    self.ptr->perspectiveRemoveIgnoredMarginPerCell = value;
}

double cv_aruco_detectorParameters_get_perspectiveRemoveIgnoredMarginPerCell(
    ArucoDetectorParams self
) {
    return self.ptr->perspectiveRemoveIgnoredMarginPerCell;
}

void cv_aruco_detectorParameters_set_maxErroneousBitsInBorderRate(
    ArucoDetectorParams self, double value
) {
    self.ptr->maxErroneousBitsInBorderRate = value;
}

double cv_aruco_detectorParameters_get_maxErroneousBitsInBorderRate(ArucoDetectorParams self) {
    return self.ptr->maxErroneousBitsInBorderRate;
}

void cv_aruco_detectorParameters_set_minOtsuStdDev(ArucoDetectorParams self, double value) {
    self.ptr->minOtsuStdDev = value;
}

double cv_aruco_detectorParameters_get_minOtsuStdDev(ArucoDetectorParams self) {
    return self.ptr->minOtsuStdDev;
}

void cv_aruco_detectorParameters_set_errorCorrectionRate(ArucoDetectorParams self, double value) {
    self.ptr->errorCorrectionRate = value;
}

double cv_aruco_detectorParameters_get_errorCorrectionRate(ArucoDetectorParams self) {
    return self.ptr->errorCorrectionRate;
}

void cv_aruco_detectorParameters_set_aprilTagQuadDecimate(ArucoDetectorParams self, float value) {
    self.ptr->aprilTagQuadDecimate = value;
}

float cv_aruco_detectorParameters_get_aprilTagQuadDecimate(ArucoDetectorParams self) {
    return self.ptr->aprilTagQuadDecimate;
}

void cv_aruco_detectorParameters_set_aprilTagQuadSigma(ArucoDetectorParams self, float value) {
    self.ptr->aprilTagQuadSigma = value;
}

float cv_aruco_detectorParameters_get_aprilTagQuadSigma(ArucoDetectorParams self) {
    return self.ptr->aprilTagQuadSigma;
}

void cv_aruco_detectorParameters_set_aprilTagMinClusterPixels(ArucoDetectorParams self, int value) {
    self.ptr->aprilTagMinClusterPixels = value;
}

int cv_aruco_detectorParameters_get_aprilTagMinClusterPixels(ArucoDetectorParams self) {
    return self.ptr->aprilTagMinClusterPixels;
}

void cv_aruco_detectorParameters_set_aprilTagMaxNmaxima(ArucoDetectorParams self, int value) {
    self.ptr->aprilTagMaxNmaxima = value;
}

int cv_aruco_detectorParameters_get_aprilTagMaxNmaxima(ArucoDetectorParams self) {
    return self.ptr->aprilTagMaxNmaxima;
}

void cv_aruco_detectorParameters_set_aprilTagCriticalRad(ArucoDetectorParams self, float value) {
    self.ptr->aprilTagCriticalRad = value;
}

float cv_aruco_detectorParameters_get_aprilTagCriticalRad(ArucoDetectorParams self) {
    return self.ptr->aprilTagCriticalRad;
}

void cv_aruco_detectorParameters_set_aprilTagMaxLineFitMse(ArucoDetectorParams self, float value) {
    self.ptr->aprilTagMaxLineFitMse = value;
}

float cv_aruco_detectorParameters_get_aprilTagMaxLineFitMse(ArucoDetectorParams self) {
    return self.ptr->aprilTagMaxLineFitMse;
}

void cv_aruco_detectorParameters_set_aprilTagMinWhiteBlackDiff(
    ArucoDetectorParams self, int value
) {
    self.ptr->aprilTagMinWhiteBlackDiff = value;
}

int cv_aruco_detectorParameters_get_aprilTagMinWhiteBlackDiff(ArucoDetectorParams self) {
    return self.ptr->aprilTagMinWhiteBlackDiff;
}

void cv_aruco_detectorParameters_set_aprilTagDeglitch(ArucoDetectorParams self, int value) {
    self.ptr->aprilTagDeglitch = value;
}

int cv_aruco_detectorParameters_get_aprilTagDeglitch(ArucoDetectorParams self) {
    return self.ptr->aprilTagDeglitch;
}

void cv_aruco_detectorParameters_set_detectInvertedMarker(ArucoDetectorParams self, bool value) {
    self.ptr->detectInvertedMarker = value;
}

bool cv_aruco_detectorParameters_get_detectInvertedMarker(ArucoDetectorParams self) {
    return self.ptr->detectInvertedMarker;
}

CvStatus* cv_aruco_Dictionary_create(ArucoDictionary* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::aruco::Dictionary();
    END_WRAP
}

CvStatus* cv_aruco_getPredefinedDictionary(int dictionaryId, ArucoDictionary* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::aruco::Dictionary(cv::aruco::getPredefinedDictionary(dictionaryId));
    END_WRAP
}

CvStatus* cv_aruco_Dictionary_create_1(
    Mat bytesList, int markerSize, int maxCorr, ArucoDictionary* rval
) {
    BEGIN_WRAP
    rval->ptr = new cv::aruco::Dictionary(CVDEREF(bytesList), markerSize, maxCorr);
    END_WRAP
}

CvStatus* cv_aruco_Dictionary_generateImageMarker(
    ArucoDictionary self, int id, int sidePixels, Mat _img, int borderBits
) {
    BEGIN_WRAP
    self.ptr->generateImageMarker(id, sidePixels, CVDEREF(_img), borderBits);
    END_WRAP
}
int cv_aruco_Dictionary_getDistanceToId(ArucoDictionary self, Mat bits, int id, bool allRotations) {
    return self.ptr->getDistanceToId(CVDEREF(bits), id, allRotations);
}
bool cv_aruco_Dictionary_identify(
    ArucoDictionary self, Mat onlyBits, int* idx, int* rotation, double maxCorrectionRate
) {
    int _idx, _rotation;
    bool rval = self.ptr->identify(CVDEREF(onlyBits), _idx, _rotation, maxCorrectionRate);
    *idx = _idx;
    *rotation = _rotation;
    return rval;
}
CvStatus* cv_aruco_Dictionary_get_bytesList(ArucoDictionary self, Mat* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Mat(self.ptr->bytesList);
    END_WRAP
}
int cv_aruco_Dictionary_get_markerSize(ArucoDictionary self) {
    return self.ptr->markerSize;
}
int cv_aruco_Dictionary_get_maxCorrectionBits(ArucoDictionary self) {
    return self.ptr->maxCorrectionBits;
}

CvStatus* cv_aruco_Dictionary_set_bytesList(ArucoDictionary self, Mat value){
    BEGIN_WRAP
    self.ptr->bytesList = CVDEREF(value);
    END_WRAP
}
void cv_aruco_Dictionary_set_markerSize(ArucoDictionary self, int value){
    self.ptr->markerSize = value;
}
void cv_aruco_Dictionary_set_maxCorrectionBits(ArucoDictionary self, int value){
    self.ptr->maxCorrectionBits = value;
}

void cv_aruco_Dictionary_close(ArucoDictionaryPtr self) {
    CVD_FREE(self);
}

CvStatus* cv_aruco_arucoDetector_create(ArucoDetector* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::aruco::ArucoDetector();
    END_WRAP
}

CvStatus* cv_aruco_arucoDetector_create_1(
    ArucoDictionary dictionary, ArucoDetectorParams params, ArucoDetector* rval
) {
    BEGIN_WRAP
    rval->ptr = new cv::aruco::ArucoDetector(CVDEREF(dictionary), CVDEREF(params));
    END_WRAP
}

void cv_aruco_arucoDetector_close(ArucoDetectorPtr self) {
    CVD_FREE(self);
}

CvStatus* cv_aruco_arucoDetector_detectMarkers(
    ArucoDetector self,
    Mat inputArr,
    VecVecPoint2f* out_markerCorners,
    VecI32* out_markerIds,
    VecVecPoint2f* out_rejectedCandidates,
    CvCallback_0 callback
) {
    BEGIN_WRAP
    self.ptr->detectMarkers(*inputArr.ptr, CVDEREF_P(out_markerCorners), CVDEREF_P(out_markerIds), CVDEREF_P(out_rejectedCandidates));
    if (callback != nullptr) {
        callback();
    }
    END_WRAP
}

CvStatus* cv_aruco_drawDetectedMarkers(
    Mat image,
    VecVecPoint2f markerCorners,
    VecI32 markerIds,
    Scalar borderColor,
    CvCallback_0 callback
) {
    BEGIN_WRAP
    cv::Scalar _borderColor =
        cv::Scalar(borderColor.val1, borderColor.val2, borderColor.val3, borderColor.val4);
    cv::aruco::drawDetectedMarkers(*image.ptr, CVDEREF(markerCorners), CVDEREF(markerIds), _borderColor);
    if (callback != nullptr) {
        callback();
    }
    END_WRAP
}

CvStatus* cv_aruco_generateImageMarker(
    int dictionaryId, int id, int sidePixels, int borderBits, Mat dst, CvCallback_0 callback
) {
    BEGIN_WRAP
    cv::aruco::Dictionary dict = cv::aruco::getPredefinedDictionary(dictionaryId);
    cv::aruco::generateImageMarker(dict, id, sidePixels, *dst.ptr, borderBits);
    if (callback != nullptr) {
        callback();
    }
    END_WRAP
}
